pkg/logql/mapper_metrics.go

```diff
@@ -0,0 +1,81 @@
+package logql
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// expression type used in metrics
+const (
+	StreamsKey = "streams"
+	MetricsKey = "metrics"
+)
+
+// parsing evaluation result used in metrics
+const (
+	SuccessKey = "success"
+	FailureKey = "failure"
+	NoopKey    = "noop"
+)
+
+// MapperMetrics is the metrics wrapper used in logql mapping (shard and range)
+type MapperMetrics struct {
+	DownstreamQueries *prometheus.CounterVec // downstream queries total, partitioned by streams/metrics
+	ParsedQueries     *prometheus.CounterVec // parsed ASTs total, partitioned by success/failure/noop
+	DownstreamFactor  prometheus.Histogram   // per request downstream factor
+}
+
+func newMapperMetrics(registerer prometheus.Registerer, mapper string) *MapperMetrics {
+	return &MapperMetrics{
+		DownstreamQueries: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace:   "loki",
+			Name:        "query_frontend_shards_total",
+			Help:        "Number of downstream queries by expression type",
+			ConstLabels: prometheus.Labels{"mapper": mapper},
+		}, []string{"type"}),
+		ParsedQueries: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace:   "loki",
+			Name:        "query_frontend_sharding_parsed_queries_total",
+			Help:        "Number of parsed queries by evaluation type",
+			ConstLabels: prometheus.Labels{"mapper": mapper},
+		}, []string{"type"}),
+		DownstreamFactor: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+			Namespace:   "loki",
+			Name:        "query_frontend_shard_factor",
+			Help:        "Number of downstream queries per request",
+			Buckets:     prometheus.LinearBuckets(0, 16, 4),
+			ConstLabels: prometheus.Labels{"mapper": mapper},
+		}),
```

Comment:
Since these metric names are now shared amongst all mappers, the names should be updated to make them more general, i.e., replacing "shards" with "downstream". However, this would result in a **breaking change**. This is my preferred option, but also don't know the impact this might have on users.
An alternative would be to remove this generalization by keeping the shard and the range mapper (and future mappers) metrics separate from their own names. Wdyt?